pyomeca/processing/rototrans.py

```diff
@@ -0,0 +1,219 @@
+    return caller(rt)
+
+
+def rototrans_from_markers(
+    caller: Callable,
+    origin: xr.DataArray,
+    axis_1: xr.DataArray,
+    axis_2: xr.DataArray,
+    axes_name: str,
+    axis_to_recalculate: str,
+) -> xr.DataArray:
+    if origin.channel.size != 1:
+        raise ValueError(
+            f"`origin` must be only one marker. You have provided {origin.channel.size} markers."
+        )
+    if axis_1.channel.size != 2:
+        raise ValueError(
+            f"`axis_1` must be two markers. You have provided {axis_1.channel.size} markers."
+        )
+    if axis_2.channel.size != 2:
+        raise ValueError(
+            f"`axis_2` must be two markers. You have provided {axis_2.channel.size} markers."
+        )
+
+    # sort the axes name - If we inverse axes_names, inverse axes as well
+    sorted_axes_name = "".join(sorted(axes_name))
+    if axes_name != sorted_axes_name:
+        axes_name = sorted_axes_name
+        axis_1, axis_2 = axis_2, axis_1
+
+    # compute vectors from markers
+    vector_1 = axis_1[:3, 1, :] - axis_1[:3, 0, :]
+    vector_2 = axis_2[:3, 1, :] - axis_2[:3, 0, :]
+
+    if origin.time.size != vector_1.time.size or origin.time.size != vector_2.time.size:
+        raise ValueError("Number of frame(s) for origin and axes must be the same")
+
+    error_msg = "Axes names should be 2 values of `x`, `y` and `z` permutations"
+
+    if axes_name[0] == "x":
+        x = vector_1
+        if axes_name[1] == "y":
+            y = vector_2
+            z = np.cross(x, y, axis=0)
+        elif axes_name[1] == "z":
+            z = vector_2
+            y = np.cross(z, x, axis=0)
+        else:
+            raise ValueError(error_msg)
+    elif axes_name[0] == "y":
+        y = vector_1
+        if axes_name[1] == "z":
+            z = vector_2
+            x = np.cross(y, z, axis=0)
+        else:
+            raise ValueError(error_msg)
+    else:
+        raise ValueError(error_msg)
+
+    if axis_to_recalculate == "x":
+        x = np.cross(y, z, axis=0)
+    elif axis_to_recalculate == "y":
+        y = np.cross(z, x, axis=0)
+    elif axis_to_recalculate == "z":
+        z = np.cross(x, y, axis=0)
+    else:
+        raise ValueError("`axis_to_recalculate must be `x`, `y` or `z`")
+
+    rt = caller(np.zeros((4, 4, origin.time.size)))
+    rt[:3, 0, :] = x / np.linalg.norm(x, axis=0)
+    rt[:3, 1, :] = y / np.linalg.norm(y, axis=0)
+    rt[:3, 2, :] = z / np.linalg.norm(z, axis=0)
+    rt.meca.translation = origin
+    return rt
+
+
+def rototrans_from_transposed_rototrans(
+    caller: Callable, rt: xr.DataArray
+) -> xr.DataArray:
+    rt_t = caller(np.zeros((4, 4, rt.time.size)))
+
+    # the rotation part is just the transposed of the rotation
+    rt_t.meca.rotation = rt.meca.rotation.transpose("col", "row", "time")
+
+    # the translation part is "- rt_t * translation"
+    rt_t.meca.translation = np.einsum(
+        "ijk,jlk->ilk", -rt_t.meca.rotation, rt.meca.translation
+    )
```

Comment:
I am still surprise that we figured this out hahaha

Reply:
This is black magic :laughing: :rofl: